### PR TITLE
[BUGFIX] Ensure that `Datasource` credentials are not persisted to Cloud/disk

### PR DESCRIPTION
**Pull request by cdkini**
Changes proposed in this pull request:
- Previously, the `Datasource`'s `config` attr contained substituted values
    - This resulted in `save_datasource` sending sensitive credentials over the wire
    - New helper and updated workflow here ensures that we always have the unsubstituted config ready for persistence

### Definition of Done
- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/data_context/config_validator/yaml_config_validator.py [hunk 1]
```diff
@@ -190,7 +190,14 @@ def test_yaml_config(  # noqa: C901 - complexity 17
...
         usage_stats_event_name: str = "data_context.test_yaml_config"
 
+        # Chetan - 20221031 -   Based on the particular object type we are attempting to instantiate,
+        # we might need the original config or the substituted config. We should standardize this
+        # such that the various private helper methods below are responsible for substitutions
+        # whenever needed.
         config = self._test_yaml_config_prepare_config(
+            yaml_config=yaml_config, usage_stats_event_name=usage_stats_event_name
+        )
+        config_with_substitutions = self._test_yaml_config_prepare_substituted_config(
```
Comment (cdkini) [inline great_expectations/data_context/config_validator/yaml_config_validator.py]: I'll try and tackle this in a subsequent PR

File: great_expectations/datasource/datasource.py [hunk 2]
```diff
@@ -196,6 +201,14 @@ def name(self):
     def config(self):
         return copy.deepcopy(self._datasource_config)
 
+    def update_config(self, config: DatasourceConfig) -> None:
+        """
+        See `BaseDatasource.update_config()` - replicated here for legacy parity.
+        """
+        substitution_serializer = DictConfigSerializer(schema=datasourceConfigSchema)
+        raw_config: dict = substitution_serializer.serialize(config)
+        self._datasource_config.update(raw_config)
```
Comment (cdkini) [inline great_expectations/datasource/datasource.py]: What's the best way to do this? Do we perhaps extend the inheritance hierarchy even further so we can place this method in a new base class? `LegacyDatasource` and `BaseDatasource` follow different inheritance hierarchies.

File: great_expectations/data_context/config_validator/yaml_config_validator.py [hunk 3]
```diff
@@ -331,9 +370,12 @@ def _test_yaml_config_prepare_config(
             )
             raise e
 
+    def _load_config_string_as_commented_map(
+        self, config_str: str, usage_stats_event_name: str
+    ) -> CommentedMap:
         try:
-            config: CommentedMap = yaml.load(config_str_with_substituted_variables)  # type: ignore[arg-type]
-            return config
+            substituted_config: CommentedMap = yaml.load(config_str)  # type: ignore[arg-type]
```
Comment (Kilo59) [inline great_expectations/data_context/config_validator/yaml_config_validator.py]: `yaml.load` doesn't accept a string??

Comment (cdkini) [reply great_expectations/data_context/config_validator/yaml_config_validator.py]: hmm not sure why we have that `type: ignore` - I'll delete and see if it works.

Is there a way to identify outdated `type: ignore` comments?

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 4]
```diff
@@ -2858,6 +2858,8 @@ def _instantiate_datasource_from_config_and_update_project_config(
                 datasource = self._instantiate_datasource_from_config(
                     config=substituted_config
                 )
+                # Sanitize instantiated object's config attr to remove any credentials with their ${VARIABLES} counterparts
+                datasource.update_config(config)
```
Comment (donaldheppner) [inline great_expectations/data_context/data_context/abstract_data_context.py]: it seems that the solution was to call an `update_config` method to restore the original config via serialization? This is strange, and the method name doesn't reflect its behavior.

I'd rather preserve the original configuration and do variable substitution on a property retrieval (ie: substitute on read, and only when the substituted config is required) rather than add extra logic to "undo" something that shouldn't have been done in the first place. Thoughts welcome!

**Review by donaldheppner — CHANGES_REQUESTED** [review]

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 4 above)
Comment (cdkini) [reply great_expectations/data_context/data_context/abstract_data_context.py]: hmm that's a very fair point - let me dig in and see how that could be done

File: great_expectations/data_context/config_validator/yaml_config_validator.py (hunk 3 above)
Comment (Kilo59) [reply great_expectations/data_context/config_validator/yaml_config_validator.py]: Yes, there is. I thought I had enabled it but I was wrong. PR up shortly to add this.
https://mypy.readthedocs.io/en/stable/config_file.html#confval-warn_unused_ignores

#6270

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 4 above)
Comment (Kilo59) [reply great_expectations/data_context/data_context/abstract_data_context.py]: A dedicated `SecretString` type whose `__str__` returns a masked value could help us here.

`pydantic` has something like this for example. But we could implement our own.
https://pydantic-docs.helpmanual.io/usage/types/#secret-types

https://dusted.codes/the-type-system-is-a-programmers-best-friend

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 5]
```diff
@@ -2776,25 +2785,27 @@ def _build_datasource_from_config(self, config: DatasourceConfig) -> Datasource:
         """
         # We convert from the type back to a dictionary for purposes of instantiation
         serializer = DictConfigSerializer(schema=datasourceConfigSchema)
-        config_dict: dict = serializer.serialize(config)
+        raw_config_dict: dict = serializer.serialize(raw_config)
+        substituted_config_dict: dict = serializer.serialize(substituted_config)
+        substituted_config_dict["raw_config"] = raw_config_dict
```
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py]: Should `DatasourceConfig` and `DatasourceConfigSchema` be updated to reflect these new changes?

File: great_expectations/datasource/new_datasource.py [hunk 6]
```diff
@@ -444,6 +445,7 @@ def __init__(
         self._datasource_config.update(
             {"data_connectors": copy.deepcopy(data_connectors)}
         )
+        self._raw_config = raw_config or self._datasource_config
```
Comment (cdkini) [inline great_expectations/datasource/new_datasource.py]: If no `raw_config` is provided through the constructor, we leverage the default config.

In all supported workflows through the `DataContext`, we will have a `raw_config`. Only if the user decides to instantiate a standalone `Datasource` will they run into this situation.

**Review by donaldheppner — APPROVED** [review]

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 5 above)
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py]: Should `DatasourceConfig` and `DatasourceConfigSchema` be updated to reflect this new config attr? `self.config` is not included in the Marshmallow schema but this is slightly different in that is it a constructor arg?

File: great_expectations/datasource/new_datasource.py [hunk 7]
```diff
@@ -59,14 +61,16 @@ def __init__(
...
-                "name": name,
-            }
         except Exception as e:
             raise ge_exceptions.ExecutionEngineError(message=str(e))
 
+        self._datasource_config: dict = {
+            "execution_engine": execution_engine,
+            "id": id,
+            "name": name,
+        }
```
Comment (cdkini) [inline great_expectations/datasource/new_datasource.py:69]: The `try/except` was for the `ExecutionEngine` creation - no need to put this `self._datasource_config` creation in the same block.

File: great_expectations/datasource/new_datasource.py [hunk 8]
```diff
@@ -59,14 +61,16 @@ def __init__(
...
+        self._raw_config = raw_config or self._datasource_config
```
Comment (cdkini) [inline great_expectations/datasource/new_datasource.py]: In the case a user instantiates a `Datasource` outside our desired context flow, we need to ensure that their `raw_config` is populated if they want to persist.

Example:
```python
datasource = Datasource(...)
context.save_datasource(datasource)
```

What if I include creds in my naked datasource? Do we allow users to leak values or should be raise warnings?

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 9]
```diff
@@ -2730,8 +2737,10 @@ def _init_datasources(self) -> None:
                 config_dict = dict(datasourceConfigSchema.dump(config))
                 datasource_config = datasourceConfigSchema.load(config_dict)
                 datasource_config.name = datasource_name
+                # Our overall config has already undergone variable substituted here so
+                # there is no difference between raw_config and substituted_config arg values
                 datasource = self._instantiate_datasource_from_config(
-                    config=datasource_config
+                    raw_config=datasource_config, substituted_config=datasource_config
```
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py]: I have a feeling this may need to change - if we look above, we substitute our ENTIRE project config (datasources and all). That means at this point, we don't have access to our raw configs.

This is used to instantiate datasources and populate the cache but might this cause issues down the line? Say I want to update an existing datasource in the cache and persist it; will I possibly leak credentials here?

I think I'll address this in a follow-up PR - the nature of the PR will be ensuring we have both `raw_config` and `substituted_config` here so we avoid problems downstream.

Comment (cdkini) [reply great_expectations/data_context/data_context/abstract_data_context.py]: Or maybe I do that in this PR idk

File: great_expectations/datasource/new_datasource.py (hunk 6 above)
Comment (anthonyburdi) [reply great_expectations/datasource/new_datasource.py]: I thought better here to return an empty config or `None` since there is no raw config in that case. That way `raw_config` is safe from any unwanted substituted values.

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 5 above)
Comment (anthonyburdi) [reply great_expectations/data_context/data_context/abstract_data_context.py]: Good question. I think so. It would have to be optional to not break existing code.

Comment (anthonyburdi) [reply great_expectations/data_context/data_context/abstract_data_context.py]: Good question. I think so. It would have to be optional to not break existing code.

File: great_expectations/datasource/new_datasource.py (hunk 8 above)
Comment (anthonyburdi) [reply great_expectations/datasource/new_datasource.py]: This is a good question. I'm not sure on this one - would be good to raise with the team.

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 9 above)
Comment (cdkini) [reply great_expectations/data_context/data_context/abstract_data_context.py]: just did it here nevermind!

File: great_expectations/datasource/new_datasource.py (hunk 6 above)
Comment (cdkini) [reply great_expectations/datasource/new_datasource.py]: X-posting from Slack:

You had noted that if omitted, raw_config should result in self._raw_config = None . This differs from my implementation, which defaults to self._raw_config = self._datasource_config).

Let’s evaluate the following example with both approaches:
```python
datasource = Datasource(...) # Assume the user is passing sensitive creds
context.save_datasource(datasource)
```

By setting self._raw_config = None, the save_datasource method will break (as the method relies on the value of self.raw_config). By setting self._raw_config = self._datasource_config, we allow the user to make the mistake of sending creds over the wire or writing them to disk.

Is there a happy middle ground? I think we need to support the example workflow I’ve noted above since that’s what is used in the Cloud snippet.

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 10]
```diff
@@ -2717,21 +2728,31 @@ def _initialize_usage_statistics(
...
+        config: DataContextConfig = self.config
         datasources: Dict[str, DatasourceConfig] = cast(
             Dict[str, DatasourceConfig], config.datasources
         )
 
+        substitutions = self._determine_substitutions()
+
         for datasource_name, datasource_config in datasources.items():
             try:
                 config = copy.deepcopy(datasource_config)  # type: ignore[assignment]
```
Comment (Kilo59) [inline great_expectations/data_context/data_context/abstract_data_context.py:2739]: ## Non-blocking

Why are we ignoring this?

File: great_expectations/datasource/new_datasource.py [hunk 11]
```diff
@@ -29,10 +29,11 @@ class BaseDatasource:
     def __init__(
         self,
         name: str,
-        execution_engine=None,
+        execution_engine: Optional[dict] = None,
```
Comment (Kilo59) [inline great_expectations/datasource/new_datasource.py:32]: ## non-blocking

Does the execution engine have an expected shape?
Can we define a `TypeDict` for it?

**Review by Kilo59 — APPROVED** [review]

File: great_expectations/datasource/new_datasource.py [hunk 12]
```diff
@@ -411,11 +423,12 @@ class Datasource(BaseDatasource):
     def __init__(
         self,
         name: str,
-        execution_engine=None,
-        data_connectors=None,
+        execution_engine: Optional[dict] = None,
+        data_connectors: Optional[dict] = None,
```
Comment (Kilo59) [inline great_expectations/datasource/new_datasource.py:421]: ## Non-blocking

Same comment as elsewhere. If there is a predictable structure to these `dict` can we define a `TypeDict` for them.

**Review by anthonyburdi — APPROVED** [review]

File: tests/data_context/abstract_data_context/test_abstract_data_context_datasource_crud.py [hunk 13]
```diff
@@ -24,20 +24,21 @@ def __init__(self):
 
 
 class FakeAbstractDataContext(AbstractDataContext):
```
Comment (Kilo59) [inline tests/data_context/abstract_data_context/test_abstract_data_context_datasource_crud.py:26]: ## Nit

Not part of the PR but this isn't a "Fake ABC" it's a "fake" concrete class right?
```suggestion
class FakeDataContext(AbstractDataContext):
```

File: great_expectations/datasource/new_datasource.py (hunk 11 above)
Comment (cdkini) [reply great_expectations/datasource/new_datasource.py:32]: as this is pretty high priority, going to defer until later. But good call!

File: great_expectations/datasource/new_datasource.py (hunk 12 above)
Comment (cdkini) [reply great_expectations/datasource/new_datasource.py:421]: Same note!

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 10 above)
Comment (cdkini) [reply great_expectations/data_context/data_context/abstract_data_context.py:2739]: I'm not sure haha let's see if `mypy` complains

**Review by anthonyburdi — APPROVED** [review]
